nodestream/pipeline/extractors/stores/aws/s3_extractor.py

```diff
@@ -1,91 +1,118 @@
-        self.logger = getLogger(__name__)
+        self.archive_dir = archive_dir
+        self.object_format = object_format
+        self.prefix = prefix or ""
 
-    @contextmanager
-    def get_object_as_tempfile(self, key: str):
-        streaming_body = self.s3_client.get_object(Bucket=self.bucket, Key=key)["Body"]
-        file = IngestibleFile.from_file_pointer_and_suffixes(
-            streaming_body, Path(key).suffixes, lambda: self.archive_s3_object(key)
-        )
-        yield file
-
-    def archive_s3_object(self, key: str):
-        if self.archive_dir:
-            self.logger.info("Archiving S3 Object", extra=dict(key=key))
-            filename = Path(key).name
-            self.s3_client.copy(
-                Bucket=self.bucket,
-                Key=f"{self.archive_dir}/{filename}",
-                CopySource={"Bucket": self.bucket, "Key": key},
-            )
-            self.s3_client.delete_object(Bucket=self.bucket, Key=key)
-
-    @contextmanager
-    def get_object_as_file(
-        self, key: str
-    ) -> Generator[SupportedFileFormat, None, None]:
-        with self.get_object_as_tempfile(key) as temp_file:
-            with SupportedFileFormat.open(temp_file, self.object_format) as file_format:
-                yield file_format
-
-    def is_object_in_archive(self, key: str) -> bool:
-        if self.archive_dir:
-            return key.startswith(self.archive_dir)
-        return False
-
-    def find_keys_in_bucket(self) -> list[str]:
-        # Returns all keys in the bucket that are not in the archive dir and have the prefix.
+    def object_is_in_archive(self, key: str) -> bool:
+        return key.startswith(self.archive_dir) if self.archive_dir else False
+
+    def find_keys_in_bucket(self) -> Iterable[str]:
+        # Returns all keys in the bucket that are not in the archive dir
+        # and have the prefix.
         paginator = self.s3_client.get_paginator("list_objects_v2")
         page_iterator = paginator.paginate(Bucket=self.bucket, Prefix=self.prefix)
         for page in page_iterator:
             keys = (obj["Key"] for obj in page.get("Contents", []))
-            yield from filter(lambda k: not self.is_object_in_archive(k), keys)
+            yield from filter(lambda k: not self.object_is_in_archive(k), keys)
 
-    async def extract_records(self) -> AsyncGenerator[Any, Any]:
+    async def get_files(self):
         for key in self.find_keys_in_bucket():
-            try:
-                with self.get_object_as_file(key) as records:
-                    for record in records.read_file():
-                        yield record
-            finally:
-                self.archive_s3_object(key)
+            yield S3File(
+                key=key,
+                s3_client=self.s3_client,
+                bucket=self.bucket,
+                archive_dir=self.archive_dir,
+                object_format=self.object_format,
+            )
+
+
+class S3Extractor(UnifiedFileExtractor):
```

Comment:
Without this class specifying the file source is to be an S3FileSource, how would we do:

UnifiedFileExtractor.from_file_data(**s3FileSourceParams) in the future?